lib/serializer/json/aws_iot_serializer_json_encoder.c

```diff
@@ -0,0 +1,691 @@
+            pContainer->pBuffer[ pContainer->offset++ ] = _JSON_ARRAY_START_CHAR;
+            break;
+        case AWS_IOT_SERIALIZER_SCALAR_TEXT_STRING:
+            pString = ( char * ) ( pScalarData->value.pString );
+            stringLength = ( pScalarData->value.stringLength == 0 ) ? strlen( pString ): pScalarData->value.stringLength;
+            _appendTextString( pContainer, pString , stringLength );
+            break;
+        case AWS_IOT_SERIALIZER_SCALAR_BYTE_STRING:
+            _appendByteString( pContainer, pScalarData->value.pString, pScalarData->value.stringLength );
+            break;
+        case AWS_IOT_SERIALIZER_SCALAR_SIGNED_INT:
+            _appendInteger( pContainer, pScalarData->value.signedInt );
+            break;
+        case AWS_IOT_SERIALIZER_SCALAR_BOOL:
+            _appendBoolean( pContainer, pScalarData->value.booleanValue );
+            break;
+        case AWS_IOT_SERIALIZER_SCALAR_NULL:
+            strncpy( ( char* ) _jsonContainerPointer( pContainer ), _JSON_NULL_VALUE, _JSON_NULL_VALUE_LENGTH );
+            break;
+        default:
+            break;
+    }
+}
+
+static void _appendJsonValue(_jsonContainer_t* pContainer,  AwsIotSerializerDataType_t xType, AwsIotSerializerScalarData_t* pxScalarData )
+{
+    if( !pContainer->isEmpty )
+    {
+        pContainer->pBuffer[ pContainer->offset++ ] = _JSON_VALUE_SEPARATOR;
+    }
+    _appendData( pContainer, xType, pxScalarData );
+}
+
+static void _appendJsonKeyValuePair(
+        _jsonContainer_t* pContainer,
+        const char * pcKey, size_t keyLength,
+        AwsIotSerializerDataType_t xValType,
+        AwsIotSerializerScalarData_t* pxScalarValue )
+{
+    if( !pContainer->isEmpty )
+    {
+        pContainer->pBuffer[ pContainer->offset++ ] = _JSON_VALUE_SEPARATOR;
+    }
+    _appendTextString( pContainer, pcKey, keyLength );
+    pContainer->pBuffer[ pContainer->offset++ ] = _JSON_KEY_VALUE_PAIR_SEPARATOR;
+    _appendData( pContainer, xValType, pxScalarValue );
+}
+
+static AwsIotSerializerError_t _init( AwsIotSerializerEncoderObject_t * pEncoderObject,
+                                      uint8_t * pDataBuffer,
+                                      size_t maxSize )
+{
+
+    _jsonContainer_t * pContainer;
+
+    /* Create a new JSON container */
+    pContainer = pvPortMalloc( sizeof( _jsonContainer_t ) );
+    if( pContainer == NULL )
+    {
+        return AWS_IOT_SERIALIZER_OUT_OF_MEMORY;
+    }
+
+    pContainer->pBuffer = pDataBuffer;
+    pContainer->maxLength = ( pDataBuffer != NULL ) ? maxSize : 0;
+    pContainer->remainingLength = pContainer->maxLength;
+    pContainer->offset = 0;
+    pContainer->overflowLength = 0;
+    pContainer->isEmpty = true;
+
+    /* Set the outermost container default type as stream */
+    pEncoderObject->type = AWS_IOT_SERIALIZER_CONTAINER_STREAM;
+
+    /* Store the container pointer within the handle */
+    pEncoderObject->pHandle = ( void* ) pContainer;
```

Comment:
minor C question: is this casting for readability or it has some good effect?

Reply:
Yes this is for readability only.